evm-adapters/src/sputnik/cheatcodes/debugger.rs

```diff
@@ -0,0 +1,403 @@
+use sputnik::{
+    Capture, ExitReason, ExitSucceed, Handler, Machine, Memory, Opcode, Resolve, Runtime,
+};
+
+use ethers::types::{Address, H256};
+
+use std::{borrow::Cow, fmt::Display, rc::Rc};
+/// EVM runtime.
+///
+/// The runtime wraps an EVM `Machine` with support of return data and context.
+pub struct ForgeRuntime<'b, 'config> {
+    pub inner: &'b mut Runtime<'config>,
+    pub code: Rc<Vec<u8>>,
+}
+
+#[derive(Debug, Clone)]
+/// An arena of `DebugNode`s
+pub struct DebugArena {
+    /// The arena of nodes
+    pub arena: Vec<DebugNode>,
+    /// The entry index, denoting the first node's index in the arena
+    pub entry: usize,
+}
+
+impl Default for DebugArena {
+    fn default() -> Self {
+        DebugArena { arena: vec![Default::default()], entry: 0 }
+    }
+}
+
+impl DebugArena {
```

Comment:
If we end up doing a blogpost on Debug and Tracing, it would be fun to "plot" the arena as it's being built, to see how the graph tree looks like.